faker/providers/__init__.py

```diff
@@ -284,18 +290,21 @@ def random_elements(self, elements=('a', 'b', 'c'), length=None, unique=False):
                 "Sample length cannot be longer than the number of unique elements to pick from.")
 
         if isinstance(elements, dict):
-            choices = elements.keys()
-            probabilities = elements.values()
+            if not hasattr(elements, "_key_cache"):
+                elements._key_cache = tuple(elements.keys())
+
+            choices = elements._key_cache
+            probabilities = tuple(elements.values()) if use_weighting else None
         else:
             if unique:
                 # shortcut
                 return self.generator.random.sample(elements, length)
             choices = elements
-            probabilities = [1.0 for _ in range(len(choices))]
+            probabilities = None
 
         return fn(
-            list(choices),
-            list(probabilities),
+            tuple(choices),
+            probabilities,
```

Comment:
I think this is an error:
try this:
```
subfaker = faker.Faker(locale=faker.config.DEFAULT_LOCALE)
subfaker.format('sentence', nb_words=2)

> {TypeError}'tuple' object does not support item assignment
``` 

because faker/providers/lorem/__init__.py:86
```
# here you return tuple
words = self.words(nb=nb_words, ext_word_list=ext_word_list)
# this assignments now fails
words[0] = words[0].title()
```

Reply:
cc: @prescod @fcurella

Reply:
issue for this: https://github.com/joke2k/faker/issues/1371